### PR TITLE
fix(server): Fix replication bug, add gdb opt to pytest

### PR DESCRIPTION
**Pull request by dranikpg**
Fixes bug in replica.cc

Add option to run instances inside gdb

### CODE REVIEW

File: tests/dragonfly/__init__.py [hunk 1]
```diff
@@ -6,27 +6,40 @@
...
 
     def start(self):
         arglist = DflyInstance.format_args(self.args)
 
         print(f"Starting instance on {self.port} with arguments {arglist}")
-        self.proc = subprocess.Popen([self.path, *arglist], cwd=self.cwd)
+
+        args = [self.params.path, *arglist]
+        if self.params.gdb:
+            args = ["gdb", "--ex", "r", "--args"] + args
```
Comment (dranikpg) [inline tests/dragonfly/__init__.py:36]: I'm not sure this option is worth it.

Its helpful for me, because in case of a bad signal, the whole process stops. Once brought back to foreground, gdb is directly active and can be used to get stacktrace/dump variables/stack. 

However, it doesn't allow customizing anything before start. Like selecting only a few instances, or setting breakpoints. The latter can be fixed by making the gdb parameter a string and forwarding it to gdb, so just `--gdb` becomes `--gdb="break dflycmd.cc:100; break replica.cc:100"`

File: tests/README.md [hunk 2]
```diff
@@ -33,15 +45,6 @@ Pytest will recursively search the `tests/dragonfly` directory for files matchin
 
 **Note**: When making a new directory in `tests/dragonfly` be sure to create an `__init__.py` file to avoid [name conflicts](https://docs.pytest.org/en/7.1.x/explanation/goodpractices.html#tests-outside-application-code)
 
-### Interacting with Dragonfly
-Pytest allows for parameters with a specific name to be automatically resolved through [fixtures](https://docs.pytest.org/en/7.1.x/explanation/fixtures.html) for any test function. The following fixtures are to be used to interact with Dragonfly when writing a test:
-| Name  | Type | [Scope](https://docs.pytest.org/en/7.1.x/how-to/fixtures.html?highlight=scope#scope-sharing-fixtures-across-classes-modules-packages-or-session) | Description
-| ----- | ---- | ----- | ----------- |
-| tmp_dir | [pathlib.Path](https://docs.python.org/3/library/pathlib.html) | Session | The temporary directory the Dragonfly binary will be running in. The environment variable `DRAGONFLY_TMP` is also set to this value |
```
Comment (romange) [inline tests/README.md:40]: Why did you delete it?

Comment (dranikpg) [reply tests/README.md:40]: Because I moved the section upwards to "important fixtures"

Comment (romange) [reply tests/README.md:40]: hmm, I do not see it. I do not find DRAGONFLY_TMP, for example.

**Comment by dranikpg** [comment]
https://github.com/dragonflydb/dragonfly/actions/runs/3565342736

should work now

**Comment by romange** [comment]
LGTM. For the protocol, can you tell what the bugs were?

File: src/server/server_family.cc [hunk 3]
```diff
@@ -782,26 +782,25 @@ static void RunStage(bool new_version, std::function<void(unsigned)> cb) {
...
+    ExtendFilenameWithShard(now, shard->shard_id(), &full_filename);
   }
-  abs_path /= file;  // use / operator to concatenate paths.
-  VLOG(1) << "Saving partial file to " << abs_path;
+  fs::path full_path = path / full_filename;  // use / operator to concatenate paths.
 
   // Start rdb saving.
   SaveMode mode = shard == nullptr ? SaveMode::SUMMARY : SaveMode::SINGLE_SHARD;
-  std::error_code local_ec = snapshot->Start(mode, abs_path.generic_string(), scripts);
+  error_code local_ec = snapshot->Start(mode, full_path.generic_string(), scripts);
```
[the remark this answers is not in the document]
Comment (dranikpg) [reply src/server/server_family.cc:803]: This is refactoring included. I broke it in the last PR unknowingly

**Comment by dranikpg** [comment]
1. Accessing replica's sync block in wrong scope when it already might be out of scope (stack placed). We could have solved it with using a shared_ptr, but this is also a logical error - no reason to do any more operations, even if they have no effect
2. Casting path to string on error and making resulting path building wrong for save df

**Comment by dranikpg** [comment]
The third change was increasing some general wait points in pytests, because the CI env is really slow. However, this has also the benefit of catching multi-threading bugs that are hard to reproduce on fast machines

**Review by romange — APPROVED** [review]
